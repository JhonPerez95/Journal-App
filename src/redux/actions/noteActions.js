import Swal from 'sweetalert2';

import { db } from '../../firebase/firebaseConfig';
import { types } from '../types/types';
import { loadNotes } from '../../helpers/loadNotes';

export const activeNote = (id, note) => {
  return {
    type: types.notesActive,
    payload: {
      id,
      ...note,
    },
  };
};

export const startNewNote = () => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime(),
    };

    const docRef = await db.collection(`${uid}/journal/notes`).add(newNote);

    dispatch(activeNote(docRef.id, newNote));
  };
};

export const setNotes = (notes) => ({
  type: types.notesLoad,
  payload: notes,
});

export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  };
};

export const startSaveNotes = (note) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;
    if (!note.url) {
      note.url = null;
    }
    const noteUpdate = { ...note };
    delete noteUpdate.id;

    await db.doc(`${uid}/journal/notes/${note.id}`).update(noteUpdate);
    dispatch(refreshNote(note.id, note));
    Swal.fire('Saved', note.title, 'success');
  };
};

export const refreshNote = (id, note) => {
  return {
    type: types.notesUpdate,
    payload: {
      id,
      ...note,
    },
  };
};
