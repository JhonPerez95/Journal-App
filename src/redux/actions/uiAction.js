import { types } from '../types/types';

export const actSetError = (payload) => {
  return {
    type: types.uiSetError,
    payload,
  };
};

export const actRemoveError = () => {
  return {
    type: types.uiRemoveError,
  };
};
