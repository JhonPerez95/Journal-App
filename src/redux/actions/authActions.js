import { types } from '../types/types';
import { firebase, auhtGoogleProvider } from '../../firebase/firebaseConfig';
import { actStartLoading, actFinishLoading } from './uiAction';

import Swal from 'sweetalert2';
import { notesLoguot } from './noteActions';

// Login
export const login = (uid, displayName) => {
  return {
    type: types.login,
    payload: {
      uid,
      displayName,
    },
  };
};

// Logout
export const logout = () => {
  return {
    type: types.logout,
  };
};

export const startLogout = () => {
  return async (dispatch) => {
    await firebase.auth().signOut();
    dispatch(logout());
    dispatch(notesLoguot());
  };
};

// Login with Email
export const startEmailLogin = (email, password) => {
  return (dispatch) => {
    dispatch(actStartLoading());
    return firebase
      .auth()
      .signInWithEmailAndPassword(email, password)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
        dispatch(actFinishLoading());
      })
      .catch((err) => {
        dispatch(actFinishLoading());
        Swal.fire('Error', err.message, 'error');
      });
  };
};

// Login with Google
export const startGoogleLogin = () => {
  return (dispatch) => {
    firebase
      .auth()
      .signInWithPopup(auhtGoogleProvider)
      .then(({ user }) => dispatch(login(user.uid, user.displayName)))
      .catch((err) => Swal.fire('Error', err.message, 'error'));
  };
};

// Register with FireBase
export const startEmailRegister = (email, password, name) => {
  return (dispatch) => {
    firebase
      .auth()
      .createUserWithEmailAndPassword(email, password)
      .then(async ({ user }) => {
        await user.updateProfile({ displayName: name });
        dispatch(login(user.uid, user.displayName));
      })
      .catch((err) => Swal.fire('Error', err.message, 'error'));
  };
};
