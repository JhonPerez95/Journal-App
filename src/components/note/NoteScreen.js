import React, { useEffect, useRef } from 'react';
import NotesAppBar from './NotesAppBar';
import { useSelector, useDispatch } from 'react-redux';
import useForm from '../../hook/useForm';
import { activeNote, startDeleting } from '../../redux/actions/noteActions';

const NoteScreen = () => {
  const dispatch = useDispatch();
  const { active: note } = useSelector((state) => state.notes);

  const [formValue, handleInputChange, resetForm] = useForm(note);
  const { title, body, url, id } = formValue;

  const activeId = useRef(note.id);

  // update form
  useEffect(() => {
    if (note.id !== activeId.current) {
      resetForm(note);
      activeId.current = note.id;
    }
  }, [note, resetForm]);

  // update state note.active
  useEffect(() => {
    dispatch(activeNote(formValue.id, { ...formValue }));
  }, [formValue, dispatch]);

  const hanldeDelete = () => {
    dispatch(startDeleting(id));
    // console.log('click Delete', id);
  };

  return (
    <div className="notes__main-content">
      <NotesAppBar />
      <div className="notes__content">
        <input
          type="text"
          className="notes__title-input"
          placeholder="Some awesome title"
          name="title"
          onChange={handleInputChange}
          value={title}
        />
        <textarea
          placeholder=" What happended today"
          className="notes__textarea"
          name="body"
          onChange={handleInputChange}
          value={body}
        ></textarea>
        {url && (
          <div className="notes__image">
            <img src={url} alt="img" />
          </div>
        )}
      </div>
      <button className="btn btn-danger" onClick={hanldeDelete}>
        Delete
      </button>
    </div>
  );
};

export default NoteScreen;
