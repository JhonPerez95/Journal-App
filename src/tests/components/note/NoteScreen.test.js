import '@testing-library/jest-dom';
import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store'; //ES6 modules

import { activeNote } from '../../../redux/actions/noteActions';
import NoteScreen from '../../../components/note/NoteScreen';

// Config Store Redux
const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
  auth: { uid: 'idTesting', name: 'Testing Name' },
  ui: { loading: false, msgError: null },
  notes: {
    notes: [],
    active: {
      id: 1234,
      body: 'Test body',
      title: 'test Title',
    },
  },
};
let store = mockStore(initState);
store.dispatch = jest.fn();

jest.mock('../../../redux/actions/noteActions', () => ({
  activeNote: jest.fn(),
}));

const wrapper = mount(
  <Provider store={store}>
    <NoteScreen />
  </Provider>
);

describe('Test the component <NoteScrren/>', () => {
  test('should show correctly ', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should call action activeNote', () => {
    wrapper.find('input[name="title"]').simulate('change', {
      target: {
        value: 'Title the Testing',
        name: 'title',
      },
    });
    expect(activeNote).toHaveBeenCalled();
  });
});
