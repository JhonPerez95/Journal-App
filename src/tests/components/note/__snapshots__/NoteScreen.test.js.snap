// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the component <NoteScrren/> should show correctly  1`] = `
<div
  className="notes__main-content"
>
  <div
    className="notes__appbar"
  >
    <span>
      August 16, 2020
    </span>
    <input
      id="inputSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div
      className=""
    >
      <button
        className="btn"
        onClick={[Function]}
      >
        Picture
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
        Save
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <input
      className="notes__title-input"
      name="title"
      onChange={[Function]}
      placeholder="Some awesome title"
      type="text"
      value="test Title"
    />
    <textarea
      className="notes__textarea"
      name="body"
      onChange={[Function]}
      placeholder=" What happended today"
      value="Test body"
    />
  </div>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
